packages/datadog-shimmer/src/shimmer.js

```diff
@@ -0,0 +1,137 @@
+'use strict'
+
+// Use a weak map to avoid polluting the wrapped function/method.
+const unwrappers = new WeakMap()
+
+function copyProperties (original, wrapped) {
+  Object.setPrototypeOf(wrapped, original)
+
+  const props = Object.getOwnPropertyDescriptors(original)
+  const keys = Reflect.ownKeys(props)
+
+  for (const key of keys) {
+    Object.defineProperty(wrapped, key, props[key])
+  }
+}
+
+function wrapFn (original, delegate) {
+  assertFunction(delegate)
+  assertNotClass(original) // TODO: support constructors of native classes
+
+  const shim = function shim () {
+    return delegate.apply(this, arguments)
+  }
+
+  unwrappers.set(shim, () => {
+    delegate = original
+  })
+
+  copyProperties(original, shim)
+
+  return shim
+}
+
+function wrapMethod (target, name, wrapper) {
+  assertMethod(target, name)
+  assertNotClass(target[name]) // TODO: support constructors of native classes
+  assertFunction(wrapper)
+
+  const original = target[name]
+  const wrapped = wrapper(original)
+  const descriptor = Object.getOwnPropertyDescriptor(target, name)
+
+  if (descriptor) {
+    unwrappers.set(wrapped, () => Object.defineProperty(target, name, descriptor))
+  } else { // no descriptor means original was on the prototype
+    unwrappers.set(wrapped, () => delete target[name])
+  }
+
+  Object.defineProperty(target, name, {
+    configurable: true,
+    writable: true,
+    enumerable: false,
+    ...descriptor,
+    value: wrapped
+  })
+
+  copyProperties(original, wrapped)
+
+  return target
+}
+
+function wrap (target, name, wrapper) {
+  return typeof name === 'function'
+    ? wrapFn(target, name)
+    : wrapMethod(target, name, wrapper)
+}
+
+function unwrap (target, name) {
+  if (!target) return target // no target to unwrap
+
+  const unwrapper = unwrappers.get(name ? target[name] : target)
+
+  if (!unwrapper) return target // target is already unwrapped or isn't wrapped
+
+  unwrapper()
+
+  return target
+}
+
+function massWrap (targets, names, wrapper) {
+  targets = [].concat(targets)
+  names = [].concat(names)
```

Comment:
We may have an opportunity here to impact startup performance by doing something like:

```js
function assertArray(maybeArray) {
  return Array.isArray(maybeArray) ? maybeArray : [maybeArray]
}
```

That way we avoid the unnecessary array creation and (usually quite expensive) `concat`. This might end up being a micro-optimization, but it could impact the startup-time benchmark.

Reply:
Also, realistically, do we use _all_ the codepaths here? I.e. 1-target-1-name, n-target-n-name, 1-target-n-name, n-target-1-name? Maybe we can simplify here based on our actual usage (rather than just the existing shimmer API).

Reply:
We do use all code paths. In my experience creating arrays and concatenating them is extremely cheap, so I doubt we'd really gain anything. Maybe I should add a benchmark?

Reply:
> In my experience creating arrays and concatenating them is extremely cheap, so I doubt we'd really gain anything.

In previous rounds of profiling, I've found them to be expensive. Right now it's (effectively) the same as previous code, so I'm fine with ignoring for now if you want to.

Reply:
Not a big change anyway so I just did it 👍